### PR TITLE
RavenDB-12672 - change incoming replication documents handling

### PR DESCRIPTION
**Pull request by grisha-kotler**

### CODE REVIEW

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 1]
```diff
@@ -891,6 +893,105 @@ private unsafe void ReadItemsFromSource(ref UnmanagedWriteBuffer writeBuffer, in
...
+            {
+                private readonly byte* _ptr;
+                private readonly long _allocationSize;
+                private readonly NativeMemory.ThreadStats _threadStats;
+                private long _used;
+                public long Free => _allocationSize - _used;
+
+                public Allocation(long allocationSize)
+                {
+                    _ptr = NativeMemory.AllocateMemory(allocationSize, out var threadStats);
```
Comment (maximburyak) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:966]: I'm not sure we want to use not pooled memory here always.
Maybe we better store some usage statistics and change this to some pooled solution, if we have a pathological high memory usage during replication

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 2]
```diff
@@ -683,12 +681,12 @@ public struct ReplicationItem : IDisposable
             #region Document
 
             public string Id;
-            public int Position;
             public string ChangeVector;
             public int DocumentSize;
             public string Collection;
             public long LastModifiedTicks;
             public DocumentFlags Flags;
+            public BlittableJsonReaderObject Document;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:689]: This struct is disposable. We probably need to dispose this?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 3]
```diff
@@ -891,6 +893,105 @@ private unsafe void ReadItemsFromSource(ref UnmanagedWriteBuffer writeBuffer, in
...
+
+                _maxSizeForContextUseInBytes = maxSizeForContextUse.GetValue(SizeUnit.Bytes);
+                var minSizeToNonContextAllocationInMb = PlatformDetails.Is32Bits ? 4 : 16;
+                _minSizeToAllocateNonContextUseInBytes = new Size(minSizeToNonContextAllocationInMb, SizeUnit.Megabytes).GetValue(SizeUnit.Bytes);
+            }
+
+            public byte* AllocateMemoryForDocument(int size)
+            {
+                TotalDocumentsSizeInBytes += size;
+                if (TotalDocumentsSizeInBytes < _maxSizeForContextUseInBytes)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: Should be `<=` in this case

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs [hunk 4]
```diff
@@ -891,6 +893,105 @@ private unsafe void ReadItemsFromSource(ref UnmanagedWriteBuffer writeBuffer, in
...
+                {
+                    _context.Allocator.Allocate(size, out var output);
+                    return output.Ptr;
+                }
+
+                if (_currentAllocation == null || _currentAllocation.Free < size)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:927]: `<=` ?

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:966]: This only kicks-in after (assuming default configuration) 128mb

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 3 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs]: done

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 4 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:927]: Here we don't have enough space on the currently allocated chunk (added a comment)

File: src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:689]: Added disposal for it.
Pretty sure that this is disposed during context dispose anyway.

Comment (grisha-kotler) [reply src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:689]: We do dispose it in MergedDocumentReplicationCommand

**Review by ppekrol — APPROVED** [review]

**Pull request merged**